src/libsystemd/sd-event/sd-event.c

```diff
@@ -386,7 +386,7 @@ static int source_io_register(
                 int enabled,
                 uint32_t events) {
 
-        struct epoll_event ev;
+        struct epoll_event ev = {0};
```

Comment:
hmm, why would this be any different then literally the same initialization a few lines down?

Also, epoll_ctl() doesn't really care about the padding. Isn't this something to fix in valgrind?

Reply:
also, we generally use the GNU C extension that allows us to write `= {}` rather than `{0}`

Reply:
valgrind is actually right, because in the [epoll_ctl syscall](https://github.com/torvalds/linux/blob/d1eef1c619749b2a57e514a3fa67d9a516ffa919/fs/eventpoll.c#L2095), it uses `copy_from_user` to copy the whole struct into kernel space, and thus the uninitialized bytes are leaked into kernel.

	    copy_from_user(&epds, event, sizeof(struct epoll_event)))

Reply:
but that's pretty much OK, no? its pading, so noone should care. And the kernel is more privileged anyway, so you can't really say "leak into". If we'd leak the padding into less privileged stuff this would be a problem, but the other direction?

Reply:
> also, we generally use the GNU C extension that allows us to write `= {}` rather than `= {0}`

OK, I could update the code to use `= {}`


> but that's pretty much OK, no? its pading, so noone should care. And the kernel is more privileged anyway, so you can't really say "leak into". If we'd leak the padding into less privileged stuff this would be a problem, but the other direction?

I agree it's not really a bug, especially if GCC is doing right we do not need this change after all.

But isn't it better to fix such uninitialized bytes when valgrind helps to identify the issue? (if we accept that GCC is not going to fix it soon, the bug was opened in 2016)

Reply:
> why would this be any different then literally the same initialization a few lines down?

I think that this is the biggest issue. Because of some accident of implementation, one initialization has a different effect than the other one. We shouldn't base our work-around something so flimsy.

I'd rather see the declaration moved down to the point of initialization and the cast removed:
```C
struct epoll_event ev = {
         .events = events | (enabled == SD_EVENT_ONESHOT ? EPOLLONESHOT : 0),
          .data.ptr = s,
};
```
Would this help with your issue?

Reply:
Nope, the problem is actually in GCC with bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=77992

It does not matter if we write
```
struct epoll_event ev;
ev = (struct epoll_event) {...};
```
or 
```
struct epoll_event ev = {...};
```

In both cases, GCC generates uninitialized data if the struct has padding bytes.
Example on coliru: http://coliru.stacked-crooked.com/a/a9a907bda006d9cd

Reply:
but shouldn't valgrind be fixed to not complain about non-initialized padding to be passed to kernel apis where it's known-safe?

I mean, i can see why complaining about that in write() makes sense, but i fail to see what could be wrong here with epoll. valrgind is just generating incorrect warnings.

Reply:
I agree that it's valid code and it should not cause anything wrong at runtime.
But I also appreciate Valgrind telling that there are non-initialized bytes that are accessed, which potentially may cause issues.

So I think manually initializing the padding to 0 is an acceptable workaround.